### ISSUE TITLE
Are categorical features handled by label encoding?

### ISSUE THREAD

**Issue by Hailprob** [microsoft/LightGBM#3700]
Are categorical features handled by label encoding?

**Comment by shiyu1994** [comment]
Currently LightGBM use an approach which is more flexible than label encoding. The labels in label encoding are replaced with gradients, and the counts in label encoding are replaced with hessians. And the encoding is done for each individual leaf (when finding split for a leaf, the encoding only uses data in the leaf). See https://lightgbm.readthedocs.io/en/latest/Advanced-Topics.html#categorical-feature-support for more details.
However, we found that this approach is more apt to overfitting. So we add another PR which uses static label encoding. See PR #3234 for more details.

**Comment by Hailprob** [comment]
Thank you for your reply, I'd like to know whether there's a tree or boosted tree method that can handle categorical features without transformation, just like continuous features in decision tree

**Comment by shiyu1994** [comment]
You mean treating categorical features as numerical ones? For example, if a categorical feature has categories {1, 2, 3}, do you want to treat 1, 2 and 3 as pure numbers (which can be compared by magnitude)? If so, you can set `categorical_feature=[]` in the `Dataset` constructor, if you're using python API.

**Comment by Hailprob** [comment]
No, if a categorical feature has categories {1, 2, 3}, I want the split is {1},{2,3} or {2},{1,3}and etc. Not {X<2.5},{X>=2.5}

**Comment by shiyu1994** [comment]
In that case, LightGBM's current approach exactly do this. The dynamic encoding I mentioned can find the optimal subset split in the form {2},{1,3}. The encoding guarantees to find the optimal split among all the 2^{k-1}-1 options, where k is the number of categories of that categorical feature.

**Comment by Hailprob** [comment]
Ok, I got it, thanks a lot, can  sklearn lgb API do it ?

**Comment by shiyu1994** [comment]
> Ok, I got it, thanks a lot, can sklearn lgb API do it ?

Yes. In the `fit` method there's a `categorical_feature` parameter.

**Comment by Hailprob** [comment]
Can't wait to try it and get how it split by plot_tree() method. I'll close the ISSUE if everything goes well

**Comment by Hailprob** [comment]
The last question that relate little to this issue, how can I dump the lightgbm.LGBMClassifier model
